src/Strategy/Swift/PackageSwift.hs

```diff
@@ -0,0 +1,248 @@
+  buildGraph,
+  parsePackageSwiftFile,
+  SwiftPackage (..),
+  SwiftPackageDep (..),
+  SwiftPackageGitDep (..),
+) where
+
+import Control.Applicative (Alternative ((<|>)), optional)
+import Control.Effect.Diagnostics (Diagnostics, context)
+import Control.Monad (void)
+import Control.Monad.Identity (Identity)
+import Data.Foldable (asum)
+import Data.Map.Strict qualified as Map
+import Data.Maybe (isJust)
+import Data.Text (Text)
+import Data.Void (Void)
+import DepTypes (DepType (GitType, SwiftType), Dependency (..), VerConstraint (CEq))
+import Effect.ReadFS (Has, ReadFS, readContentsParser)
+import Graphing (Graphing, directs, induceJust)
+import Path
+import Text.Megaparsec (
+  MonadParsec (takeWhile1P, try),
+  Parsec,
+  ParsecT,
+  anySingle,
+  between,
+  empty,
+  sepEndBy,
+  skipManyTill,
+ )
+import Text.Megaparsec.Char (space1)
+import Text.Megaparsec.Char.Lexer qualified as Lexer
+import Types (GraphBreadth (..))
+
+-- | Parsing
+-- *
+type Parser = Parsec Void Text
+
+sc :: Parser ()
+sc =
+  Lexer.space
+    space1
+    (Lexer.skipLineComment "//")
+    (Lexer.skipBlockComment "/*" "*/")
+
+lexeme :: Parser a -> Parser a
+lexeme = Lexer.lexeme sc
+
+symbol :: Text -> Parser Text
+symbol = Lexer.symbol sc
+
+scWOComment :: Parser ()
+scWOComment = Lexer.space space1 empty empty
+
+symbolWOComment :: Text -> Parser Text
+symbolWOComment = Lexer.symbol scWOComment
+
+betweenDoubleQuotes :: Parser a -> Parser a
+betweenDoubleQuotes = between (symbol "\"") (symbol "\"")
+
+betweenSquareBrackets :: Parser a -> Parser a
+betweenSquareBrackets = between (symbol "[") (symbol "]")
+
+betweenBrackets :: Parser a -> Parser a
+betweenBrackets = between (symbol "(") (symbol ")")
+
+maybeComma :: Parser ()
+maybeComma = void $ optional $ lexeme $ symbol ","
+
+parseQuotedText :: Parser Text
+parseQuotedText = betweenDoubleQuotes (lexeme $ takeWhile1P (Just "quoted text") (/= '"'))
+
+parseKeyValue :: Text -> Parser a -> Parser a
+parseKeyValue t parser = lexeme $ symbol (t <> ":") *> parser
+
+isEndLine :: Char -> Bool
+isEndLine '\n' = True
+isEndLine '\r' = True
+isEndLine _ = False
+
+-- | Represents https://github.com/apple/swift-package-manager/blob/main/Documentation/PackageDescription.md#methods.
+data SwiftPackage = SwiftPackage
+  { swiftToolVersion :: Text
+  , packageDependencies :: [SwiftPackageDep]
+  }
+  deriving (Show, Eq, Ord)
+
+-- | Represents https://github.com/apple/swift-package-manager/blob/main/Documentation/PackageDescription.md#package-dependency.
+data SwiftPackageDep
+  = GitSource SwiftPackageGitDep
+  | PathSource Text
+  deriving (Show, Eq, Ord)
+
+data SwiftPackageGitDep = SwiftPackageGitDep
+  { srcOf :: Text
+  , branchOf :: Maybe Text
+  , revisionOf :: Maybe Text
+  , fromOf :: Maybe Text
+  , exactOf :: Maybe Text
+  , upToNextMajorOf :: Maybe Text
+  , upToNextMinorOf :: Maybe Text
+  , closedInterval :: Maybe (Text, Text)
+  , rhsHalfOpenInterval :: Maybe (Text, Text)
+  }
+  deriving (Show, Eq, Ord)
```

Comment:
How many constraints can a single dep have?  If it's not all of these, then we should probably use a sum type instead for the constraints.

Reply:
It can only one - good catch. Fixed in latest commit.